### PR TITLE
TS-4664: Fix crash by unifying event handlers for ClientSession.

### PR DESCRIPTION
**Pull request by shinrich**
Combining the handlers for ClientSession so both IO events and plugin events can be handled.

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/434/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/327/ for details.

File: proxy/ProxyClientSession.cc [hunk 1]
```diff
@@ -142,7 +141,9 @@ ProxyClientSession::do_api_callout(TSHttpHookID id)
   this->api_current = NULL;
 
   if (this->hooks_on && this->has_hooks()) {
-    SET_HANDLER(&ProxyClientSession::state_api_callout);
+    if (!this->handler) {
```
Comment (jpeach) [inline proxy/ProxyClientSession.cc]: I'm not sure about this. The state of the handler doesn't seem to be well-defined at this point. HTTP2 sets the handler in ``start()``, for example.

File: proxy/ProxyClientSession.cc [hunk 2]
```diff
@@ -142,7 +141,9 @@ ProxyClientSession::do_api_callout(TSHttpHookID id)
   this->api_current = NULL;
 
   if (this->hooks_on && this->has_hooks()) {
```
Comment (jpeach) [inline proxy/ProxyClientSession.cc]:
```C
if (this->hooks_enabled() && this->has_hooks())
```

**Comment by jpeach** [comment]
I had to re-read this code to refresh my memory on how this works. I think that the invariant that the ``ProxyClientSession`` requires is that the API hooks are the first and last operations performed on the session. This means that the handler must be ``NULL`` when delivering the ``SSN_START`` event (which is always true because this is done before ``start()``). It also means that we should not have to worry about other event handlers when delivering ``SSN_CLOSE``, because the only ``destroy()`` happens after this.

I agree that ``state_api_callout`` should not be virtual (it should be private).

File: proxy/ProxyClientSession.cc (hunk 1 above)
Comment (shinrich) [inline proxy/ProxyClientSession.cc]: Hmm.  It looks like Http1ClientSession::start does not set the handler.  I could add the handler check there and remove the null check here.

**Comment by shinrich** [comment]
I'm not sure how to interpret your comment.  I was seeing IO events trailing during the SSN close handling.  For example, it looks like I could call the clearing do_io_write in Http1ClientSession::destroy() instead of Http1ClientSession::free().

However, it seems that having a single handler deal with both sets of events (I/O and plugin) is more resilient.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/448/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/342/ for details.

File: configure.ac [hunk 3]
```diff
@@ -49,7 +49,7 @@ AM_INIT_AUTOMAKE([-Wall -Werror -Wno-portability tar-ustar foreign no-installinf
 AM_MAINTAINER_MODE([enable])
 
 # Enable a recursive "tidy" rule for clang-tidy.
-AM_EXTRA_RECURSIVE_TARGETS([tidy])
+#AM_EXTRA_RECURSIVE_TARGETS([tidy])
```
Comment (jpeach) [inline configure.ac]: Please undo this one :)

**Comment by jpeach** [comment]
Ah. The invariant is that you are not supposed to be getting IO events during session hook processing. Maybe the fact that this is happening is the root cause of the bug?

This change adds knowledge of the hook processing into more event handlers which seems worse to me. The logic now needs to be maintained and understood in more places. The original goal of the base class here was to remove the duplication of session hook processing.

File: configure.ac (hunk 3 above)
Comment (shinrich) [inline configure.ac]: Sorry snuck in.  My Centos image cannot deal with this one.

**Comment by shinrich** [comment]
We are running with this change locally.  I've made some other changes on this front too.   I'll look at adding some asserts/warning messages to see whether we are still encountering this issue.  The problem is on the closing side not the start up side.

**Comment by shinrich** [comment]
I'm actually holding off on this one.  We are running our version of 5.3.x with this change, but there were a number of other changes as well.  Not clear this one is needed.  I plan to do stress testing of 7.x this month.  I'll pull this logic forward if it is needed (or at least understand the issue better).

**Comment by shinrich** [comment]
Looks like this one isn't needed. Closing.

**Pull request closed without merging**